packages/wrangler/src/type-generation.ts

```diff
@@ -0,0 +1,171 @@
+		}
+	}
+
+	if (configToDTS.r2_buckets) {
+		for (const R2Bucket of configToDTS.r2_buckets) {
+			envTypeStructure.push(`	${R2Bucket.binding}: R2Bucket;`);
+		}
+	}
+
+	if (configToDTS.d1_databases) {
+		for (const d1 of configToDTS.d1_databases) {
+			envTypeStructure.push(`	${d1.binding}: D1Database;`);
+		}
+	}
+
+	if (configToDTS.services) {
+		for (const service of configToDTS.services) {
+			envTypeStructure.push(`	${service.binding}: Fetcher;`);
+		}
+	}
+
+	if (configToDTS.dispatch_namespaces) {
+		for (const namespace of configToDTS.dispatch_namespaces) {
+			envTypeStructure.push(`	${namespace.binding}: any;`);
+		}
+	}
+
+	if (configToDTS.logfwdr?.schema) {
+		envTypeStructure.push(`	LOGFWDR_SCHEMA: any;`);
+	}
+
+	if (configToDTS.data_blobs) {
+		for (const dataBlobs in configToDTS.data_blobs) {
+			envTypeStructure.push(`	${dataBlobs}: ArrayBuffer;`);
+		}
+	}
+
+	if (configToDTS.text_blobs) {
+		for (const textBlobs in configToDTS.text_blobs) {
+			envTypeStructure.push(`	${textBlobs}: string;`);
+		}
+	}
+
+	if (configToDTS.unsafe) {
+		for (const unsafe of configToDTS.unsafe) {
+			envTypeStructure.push(`	${unsafe.name}: any;`);
+		}
+	}
+
+	const modulesTypeStructure: string[] = [];
+	if (configToDTS.rules) {
+		const typeToTypeScriptType = {
+			Text: "string",
+			Data: "ArrayBuffer",
+			CompiledWasm: "WebAssembly.Module",
+		};
+		for (const ruleObject of configToDTS.rules) {
+			const typeScriptType =
+				typeToTypeScriptType[
+					ruleObject.type as keyof typeof typeToTypeScriptType
+				];
+			if (typeScriptType !== undefined) {
+				ruleObject.globs.forEach((glob) => {
+					modulesTypeStructure.push(`declare module "${glob}" {
+	const value: ${typeScriptType};
+	export default value;
+}`);
+				});
+			}
+		}
+	}
+
+	function writeDTSFile(
+		typesString: string[],
+		formatType: "modules" | "service-worker"
+	) {
+		const wranglerOverrideDTSPath = findUpSync("wrangler-overrides.d.ts");
+		try {
+			if (
+				wranglerOverrideDTSPath !== undefined &&
+				!fs
+					.readFileSync(wranglerOverrideDTSPath, "utf8")
+					.includes("***AUTO GENERATED BY WORKERS CLI WRANGLER***")
+			) {
+				throw new Error(
+					"A non-wrangler wrangler-overrides.d.ts already exists, please rename and try again."
+				);
+			}
+		} catch (error) {
+			if (error instanceof Error && !error.message.includes("not found")) {
+				throw error;
+			}
+		}
+
+		let combinedTypeStrings = "";
+		if (formatType === "modules") {
+			combinedTypeStrings =
+				"interface Env {" +
+				"\n" +
+				typesString.join("\n") +
+				"\n" +
+				"}" +
+				"\n" +
+				modulesTypeStructure.join("\n");
+		} else {
+			combinedTypeStrings =
+				"declare global {" +
+				"\n" +
+				typesString.join("\n") +
+				"\n" +
+				"}" +
+				"\n" +
+				modulesTypeStructure.join("\n");
+		}
+
+		if (envTypeStructure.length || modulesTypeStructure.length) {
+			fs.writeFileSync(
+				"wrangler-overrides.d.ts",
+				"// ***AUTO GENERATED BY WORKERS CLI WRANGLER***" +
```

Comment:
Maybe something like `Generated by Wrangler on ${date}`?

Reply:
Oh, I like this.